backend/auth/firebase.go

```diff
@@ -16,55 +16,92 @@ package auth
+func (auth FirebaseAuth) Authenticate(next http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(r.Context(), auth.AuthTimeout)
+		defer cancel()
+
 		idToken := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
-		userEmail, err := auth.Authenticate(r.Context(), idToken)
+		token, err := auth.FirebaseClient.VerifyIDToken(ctx, idToken)
 		if err != nil {
 			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
-			http.Error(w, err.Error(), http.StatusUnauthorized)
+			http.Error(w, "User authentication failed", http.StatusUnauthorized)
 			return
 		}
-		if getDomain(userEmail) != auth.AuthDomain {
-			http.Error(w, "You are not authorized to view this resource", http.StatusForbidden)
+		if !token.Claims["email_verified"].(bool) {
+			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
+			http.Error(w, "User authentication failed, please verify your email address", http.StatusUnauthorized)
 			return
 		}
-		next(w, r)
+
+		userEmail := token.Claims["email"].(string)
+		user := User{
+			email:  userEmail,
+			domain: getDomain(userEmail),
+		}
+
+		ctxWithUser := context.WithValue(r.Context(), "user", user)
+		next(w, r.WithContext(ctxWithUser))
 	}
 }
 
-func (auth *FirebaseAuth) Authenticate(ctx context.Context, idToken string) (userEmail string, err error) {
-	ctx, cancel := context.WithTimeout(ctx, auth.AuthTimeout)
-	defer cancel()
+func getPermissionsFromTeam(team models.Team, action string) []models.Principal {
+	var permissions []models.Principal
+	if action == ActionRead {
+		permissions = team.Permissions.Read.Allow
+	} else if action == ActionWrite {
+		permissions = team.Permissions.Write.Allow
+	}
+	return permissions
+}
 
-	token, err := auth.FirebaseClient.VerifyIDToken(ctx, idToken)
-	errorMessage := ""
-	if err != nil {
-		errorMessage = "User authentication failed "
-		return "", errors.New(errorMessage)
+func getPermissionsFromGeneral(generalPermissions models.GeneralPermissions, action string) []models.Principal {
+	var permissions []models.Principal
+	if action == ActionReadTeamList {
+		permissions = generalPermissions.ReadTeamList.Allow
+	} else if action == ActionAddTeam {
+		permissions = generalPermissions.AddTeam.Allow
 	}
-	if !token.Claims["email_verified"].(bool) {
-		errorMessage = "User authentication failed, please verify your email address"
-		return "", errors.New(errorMessage)
+	return permissions
+}
+
+func (user User) has(permission models.Principal) bool {
+	return (permission.Type == models.PrincipalTypeDomain && strings.ToLower(permission.ID) == strings.ToLower(user.domain)) ||
+		(permission.Type == models.PrincipalTypeEmail && strings.ToLower(permission.ID) == strings.ToLower(user.email))
+}
+
+func (user User) isPermitted(permissions []models.Principal) bool {
+	isPermitted := false
+	for _, permission := range permissions {
+		if user.has(permission) {
+			isPermitted = true
```

Comment:
Can just short-cut `return true` here, then `return false` at the end if none of them matched.

Reply:
Ah yes, this was a relic from when this chunk was still part of a larger function but I should have caught this.